cranelift/codegen/src/isa/riscv64/inst.isle

```diff
@@ -999,50 +988,121 @@
       ;; add low and high together.
       (result Reg (alu_add high low)))
     (value_regs result (load_u64_constant 0))))
+  
+;; Extends an integer if it is smaller than 64 bits.
+(decl ext_int_if_need (bool ValueRegs Type) ValueRegs)
+;;; For values smaller than 64 bits, we need to extend them to 64 bits
+(rule 0 (ext_int_if_need $true val (fits_in_32 (ty_int ty)))
+  (sext val ty $I64))
+(rule 0 (ext_int_if_need $false val (fits_in_32 (ty_int ty)))
+  (zext val ty $I64))
+;; If the value is larger than one machine register, we don't need to do anything
+(rule 1 (ext_int_if_need _ r $I64) r)
+(rule 2 (ext_int_if_need _ r $I128) r)
 
-(decl gen_extend (Reg bool u8 u8) Reg)
-(rule
-  (gen_extend r is_signed from_bits to_bits)
-  (let
-    ((tmp WritableReg (temp_writable_reg $I16))
-      (_ Unit (emit (MInst.Extend tmp r is_signed from_bits to_bits))))
-    tmp))
 
-;; val is_signed from_bits to_bits
-(decl lower_extend (Reg bool u8 u8) ValueRegs)
-(rule -1
-  (lower_extend r is_signed from_bits to_bits)
-  (gen_extend r is_signed from_bits to_bits))
+;; Performs a zero extension of the given value
+(decl zext (ValueRegs Type Type) ValueRegs)
+(rule (zext val from_ty to_ty) (extend val (ExtendOp.Zero) from_ty to_ty))
 
-;;;; for I128 signed extend.
-(rule 1
-  (lower_extend r $true 64 128)
-  (let
-    ((tmp Reg (alu_rrr (AluOPRRR.Slt) r (zero_reg)))
-      (high Reg (gen_extend tmp $true 1 64)))
-    (value_regs (gen_move2 r $I64 $I64) high)))
+;; Performs a signed extension of the given value
+(decl sext (ValueRegs Type Type) ValueRegs)
+(rule (sext val from_ty to_ty) (extend val (ExtendOp.Signed) from_ty to_ty))
 
-(rule
-  (lower_extend r $true from_bits 128)
-  (let
-    ((tmp Reg (gen_extend r $true from_bits 64))
-      (tmp2 Reg (alu_rrr (AluOPRRR.Slt) tmp (zero_reg)))
-      (high Reg (gen_extend tmp2 $true 1 64)))
-    (value_regs (gen_move2 tmp $I64 $I64) high)))
+(type ExtendOp
+  (enum
+    (Zero)
+    (Signed)))
+
+;; Performs either a sign or zero extension of the given value
+(decl extend (ValueRegs ExtendOp Type Type) ValueRegs)
+
+;;; Generic Rules Extending to I64
+(decl extend_shift_op (ExtendOp) AluOPRRI)
+(rule (extend_shift_op (ExtendOp.Zero)) (AluOPRRI.Srli))
+(rule (extend_shift_op (ExtendOp.Signed)) (AluOPRRI.Srai))
```

Comment:
Having something like a match would be really nice! Or at least being able to scope these rules so that they don't accidentally get used by other rules.